core/src/main/java/org/mule/runtime/core/util/ExceptionUtils.java

```diff
@@ -255,6 +265,87 @@ public static Event createErrorEvent(Event currentEvent, Object annotatedObject,
+    Optional<Exception> rootExceptionOptional =
+        findRootExceptionForErrorHandling(exception, errorTypeLocator, errorTypeRepository);
+
+    Exception rootException;
+    if (rootExceptionOptional.isPresent()) {
+      if (logger.isDebugEnabled()) {
+        logger.debug("discarding exception that is wrapping the original error", exception);
+      }
+      if (rootExceptionOptional.get() instanceof MessagingException
+          && ((MessagingException) rootExceptionOptional.get()).getEvent().getError().isPresent()) {
+        return (MessagingException) rootExceptionOptional.get();
+      }
+      rootException = rootExceptionOptional.get();
+    } else {
+      rootException = exception;
+    }
+
+    Processor failing = exception.getFailingMessageProcessor();
+    if (failing == null && rootException instanceof MessagingException) {
+      failing = ((MessagingException) rootException).getFailingMessageProcessor();
+    }
+
+    if (failing == null) {
+      failing = processor;
+      exception = new MessagingException(createStaticMessage(rootException.getMessage()), exception.getEvent(),
+                                         rootException instanceof MessagingException ? rootException.getCause() : rootException,
+                                         processor);
+    }
+    exception
+        .setProcessedEvent(createErrorEvent(exception.getEvent(), processor, exception, errorTypeLocator));
+    return putContext(exception, failing, exception.getEvent(), flowConstruct, muleContext);
+  }
+
+  /**
+   * Searches for the root {@link MuleException} to use to generate the {@link Error} inside the {@link Event}.
+   * <p>
+   * If such exception exists, then it's because the exception is wrapping an exception that already has an error. For instance, a
+   * streaming error
+   *
+   * @param exception the exception to search in all it's causes for a {@link MessagingException} with an {@link Error}
+   * @return the found exception or empty.
+   */
+  private static Optional<Exception> findRootExceptionForErrorHandling(Exception exception, ErrorTypeLocator errorTypeLocator,
+                                                                       ErrorTypeRepository errorTypeRepository) {
+    List<Throwable> causesAsList = getExceptionsAsList(exception);
+    for (Throwable cause : causesAsList) {
+      if (cause instanceof TypedException) {
+        return of((Exception) cause);
+      }
+      if (cause instanceof MuleException || cause instanceof MuleRuntimeException) {
+        ErrorType errorType = errorTypeLocator.lookupErrorType(cause);
+        if (!errorType.equals(errorTypeRepository.getErrorType(buildFromStringRepresentation(UNKNOWN_ERROR_IDENTIFIER)))) {
+          return of((Exception) cause);
+        }
+      }
+    }
+    Optional<MessagingException> messagingExceptionOptional = getDeepestExceptionType(MessagingException.class, exception);
```

Comment:
I'm not sure why the order here is not inverted, shouldn't we check for an existing error first and then try to resolve it?